File: backend/core/cutter.py
"""FFmpeg video trimming and aspect-ratio cropping engine."""

from __future__ import annotations

import os
import re
import shutil
import subprocess

PORTRAIT = "portrait"
LANDSCAPE = "landscape"
ORIGINAL = "original"

PORTRAIT_FILTER = "crop=min(iw\\,ih*9/16):ih:(iw-min(iw\\,ih*9/16))/2:0"
LANDSCAPE_FILTER = "crop=iw:min(ih\\,iw*9/16):0:(ih-min(ih\\,iw*9/16))/2"

VIDEO_EXTENSIONS = {".mp4", ".mov", ".mkv", ".webm", ".avi", ".m4v", ".ts"}


class CutterError(Exception):
    """Raised when FFmpeg is unavailable or a cut fails."""


def crop_filter_for(orientation: str) -> str | None:
    """Return the crop filter for ``orientation``, or None for no crop.

    - ``portrait`` crops to a 9:16 vertical window.
    - ``landscape`` crops to a 16:9 horizontal window.
    - ``original`` (or None/empty) leaves the source ratio untouched.
    """
    if orientation in (None, "", PORTRAIT):
        return PORTRAIT_FILTER
    if orientation == LANDSCAPE:
        return LANDSCAPE_FILTER
    if orientation == ORIGINAL:
        return None
    raise CutterError(f"Unknown orientation: {orientation!r}")


def verify_ffmpeg() -> str:
    """Return the path to ffmpeg, or raise CutterError if not found."""
    path = shutil.which("ffmpeg")
    if not path:
        raise CutterError(
            "ffmpeg was not found on the system PATH. Install ffmpeg and add it to PATH."
        )
    return path


def slugify(title: str) -> str:
    """Turn a clip title into a filesystem-safe filename slug."""
    slug = re.sub(r"[^A-Za-z0-9]+", "-", title.strip()).lower().strip("-")
    return slug or "clip"


def _escape_filter_path(path: str) -> str:
    """Escape a filesystem path for use inside an ffmpeg filter argument.

    ffmpeg filtergraph syntax treats ``:``, ``\\``, ``'`` and ``[``/``]`` as
    special; the subtitles filter's own docs show colons in Windows-style
    drive paths escaped as ``\\:``. Backslashes must be escaped first so we
    don't double-escape the backslashes we just inserted.
    """
    return path.replace("\\", "\\\\").replace(":", "\\:").replace("'", "\\'")


def subtitles_filter_for(ass_path: str, fonts_dir: str | None = None) -> str:
    """Build the ``subtitles=`` filter fragment that burns an ASS file in."""
    escaped = _escape_filter_path(ass_path)
    parts = [f"subtitles={escaped}"]
    if fonts_dir:
        parts.append(f"fontsdir={_escape_filter_path(fonts_dir)}")
    return ":".join(parts)


def build_command(
    src: str,
    start: float,
    end: float,
    title: str,
    out_dir: str,
    index: int,
    orientation: str = PORTRAIT,
    subtitles_path: str | None = None,
    fonts_dir: str | None = None,
) -> list[str]:
    """Build the FFmpeg command that trims a clip to the chosen ratio.

    - ``-ss`` appears before ``-i`` for fast seeking, so we pass ``-t``
      (duration) rather than ``-to``.
    - ``orientation`` selects the crop filter: portrait (9:16), landscape
      (16:9), or original (no crop).
    - ``subtitles_path`` (optional): an ASS subtitle file to burn in via
      libass. Combined with the crop filter in one ``-vf`` chain so
      captions are positioned against the already-cropped frame.
    """
    duration = max(end - start, 0.1)
    filename = f"{slugify(title)}_{index:02d}.mp4"
    out_path = os.path.join(out_dir, filename)

    command = [
        "ffmpeg",
        "-y",
        "-hide_banner",
        "-loglevel",
        "error",
        "-ss",
        f"{start:.2f}",
        "-i",
        src,
        "-t",
        f"{duration:.2f}",
    ]

    filters = []
    crop_filter = crop_filter_for(orientation)
    if crop_filter:
        filters.append(crop_filter)
    if subtitles_path:
        filters.append(subtitles_filter_for(subtitles_path, fonts_dir))
    if filters:
        command += ["-vf", ",".join(filters)]

    command += [
        "-c:v",
        "libx264",
        "-preset",
        "veryfast",
        "-crf",
        "20",
        "-c:a",
        "aac",
        "-b:a",
        "128k",
        "-movflags",
        "+faststart",
        out_path,
    ]
    return command


def cut_clip(
    src: str,
    start: float,
    end: float,
    title: str,
    out_dir: str,
    index: int,
    orientation: str = PORTRAIT,
    subtitles_path: str | None = None,
    fonts_dir: str | None = None,
) -> str:
    """Cut a single clip to out_dir and return the output file path.

    ``subtitles_path`` (optional) is an ASS file burned in via libass;
    ``fonts_dir`` (optional) is a font directory passed to libass.
    """
    if not os.path.isfile(src):
        raise CutterError(f"Source video not found: {src}")

    verify_ffmpeg()

    os.makedirs(out_dir, exist_ok=True)

    command = build_command(
        src,
        start,
        end,
        title,
        out_dir,
        index,
        orientation,
        subtitles_path=subtitles_path,
        fonts_dir=fonts_dir,
    )
    result = subprocess.run(command, capture_output=True, text=True)

    if result.returncode != 0:
        stderr_tail = (result.stderr or result.stdout or "").strip().splitlines()
        detail = stderr_tail[-1] if stderr_tail else "unknown error"
        raise CutterError(f"FFmpeg failed for clip {title!r}: {detail}")

    return command[-1]

File: electron/src/services/cutter.ts
import path from "node:path";
import fs from "node:fs";
import { spawn } from "node:child_process";
import { ffmpegPath } from "./bin.js";

export const PORTRAIT = "portrait";
export const LANDSCAPE = "landscape";
export const ORIGINAL = "original";

const PORTRAIT_FILTER = "crop=min(iw\\,ih*9/16):ih:(iw-min(iw\\,ih*9/16))/2:0";
const LANDSCAPE_FILTER = "crop=iw:min(ih\\,iw*9/16):0:(ih-min(ih\\,iw*9/16))/2";

export class CutterError extends Error {}

export function cropFilterFor(o: string): string | null {
  if (!o || o === PORTRAIT) return PORTRAIT_FILTER;
  if (o === LANDSCAPE) return LANDSCAPE_FILTER;
  if (o === ORIGINAL) return null;
  throw new CutterError(`Unknown orientation: ${o}`);
}

export function faceAwareCropFilter(crop: { x: number; y: number; w: number; h: number }): string {
  // crop is already evenDown + clamped inside source; use literal xywh
  return `crop=${crop.w}:${crop.h}:${crop.x}:${crop.y}`;
}



function escapeFilterPath(p: string): string {
  return p.replace(/\\/g, "\\\\").replace(/:/g, "\\:").replace(/'/g, "\\'");
}

function subtitlesFilterFor(assPath: string, fontsDir?: string): string {
  const e = escapeFilterPath(assPath);
  // Wrap in single quotes to handle Windows paths with spaces (e.g. C:\Users\AL EL FAMILY\...)
  // FFmpeg libass expects the filename to be quoted when it contains spaces or special chars
  return fontsDir ? `subtitles='${e}':fontsdir='${escapeFilterPath(fontsDir)}'` : `subtitles='${e}'`;
}

function scaleFilter(maxRes?: number | null): string | null {
  if (!maxRes) return null;
  return `scale=min(iw\\,${maxRes}):min(ih\\,${maxRes}):force_original_aspect_ratio=decrease`;
}

export function slugify(title: string): string {
  return title.trim().replace(/[^A-Za-z0-9]+/g, "-").toLowerCase().replace(/^-|-$/g, "") || "clip";
}

export function buildCommand(
  src: string,
  start: number,
  end: number,
  title: string,
  outDir: string,
  index: number,
  orientation = PORTRAIT,
  subtitlesPath?: string | null,
  fontsDir?: string | null,
  maxResolution?: number | null,
  faceCrop?: { x: number; y: number; w: number; h: number } | null
): string[] {
  const duration = Math.max(end - start, 0.1);
  const outPath = path.join(outDir, `${slugify(title)}_${String(index).padStart(2, "0")}.mp4`);
  const cmd: string[] = [ffmpegPath(), "-y", "-hide_banner", "-loglevel", "error", "-ss", start.toFixed(2), "-i", src, "-t", duration.toFixed(2)];
  const filters: string[] = [];
  const crop = faceCrop ? faceAwareCropFilter(faceCrop) : cropFilterFor(orientation);
  if (crop) filters.push(crop);
  const scale = scaleFilter(maxResolution);
  if (scale) filters.push(scale);
  if (subtitlesPath) filters.push(subtitlesFilterFor(subtitlesPath, fontsDir ?? undefined));
  if (filters.length) cmd.push("-vf", filters.join(","));
  cmd.push("-c:v", "libx264", "-preset", "veryfast", "-crf", "20", "-c:a", "aac", "-b:a", "128k", "-movflags", "+faststart", outPath);
  return cmd;
}

export function cutClip(
  src: string,
  start: number,
  end: number,
  title: string,
  outDir: string,
  index: number,
  orientation = PORTRAIT,
  subtitlesPath?: string | null,
  fontsDir?: string | null,
  maxResolution?: number | null,
  faceCrop?: { x: number; y: number; w: number; h: number } | null
): Promise<string> {
  if (!fs.existsSync(src)) return Promise.reject(new CutterError(`Source not found: ${src}`));
  fs.mkdirSync(outDir, { recursive: true });
  const cmd = buildCommand(src, start, end, title, outDir, index, orientation, subtitlesPath, fontsDir, maxResolution, faceCrop ?? null);
  const outPath = cmd[cmd.length - 1];
  console.log(`[cutter] ffmpeg bin=${cmd[0]} exists=${fs.existsSync(String(cmd[0]))} src=${src} start=${start} end=${end} out=${outPath}`);
  console.log(`[cutter] cmd: ${cmd.join(" ")}`);
  if (subtitlesPath) console.log(`[cutter] subtitlesPath=${subtitlesPath} exists=${fs.existsSync(String(subtitlesPath))} fontsDir=${fontsDir}`);
  return new Promise((resolve, reject) => {
    const [bin, ...args] = cmd;
    const p = spawn(bin, args, { stdio: "pipe" });
    let stderr = "";
    let stdout = "";
    p.stderr.on("data", (d) => (stderr += d.toString()));
    p.stdout.on("data", (d) => (stdout += d.toString()));
    p.on("error", (e) => {
      console.error(`[cutter] spawn error bin=${bin} err=${String(e)} exists=${fs.existsSync(String(bin))}`);
      reject(new CutterError(`FFmpeg spawn failed for ${title}: ${String(e)} (bin=${bin})`));
    });
    p.on("close", (code) => {
      console.log(`[cutter] close code=${code} stderr=${stderr.slice(0, 2000)} stdout=${stdout.slice(0, 500)} outExists=${fs.existsSync(outPath)}`);
      if (code === 0 && fs.existsSync(outPath)) resolve(outPath);
      else {
        const fullErr = (stderr || stdout || `exit code ${code}`).trim().slice(0, 2000) || "unknown";
        reject(new CutterError(`FFmpeg failed for ${title}: ${fullErr} (bin=${bin} code=${code})`));
      }
    });
  });
}
